diffrax/integrate.py

```diff
@@ -142,34 +180,108 @@ def body_fun(state, inplace):
                 state.made_jump,
             )
         else:
-            # TODO: double-check that these are the correct `ts_size` and
-            # `direction`.
-            history = DenseInterpolation(
+
+            min_delay = []
+            flat_delays = jtu.tree_leaves(delays)
+            for delay in flat_delays:
+                min_delay.append(delay(state.tprev, state.y, args))
+            min_delay = jnp.stack(min_delay).min()
+            implicit_step = min_delay < state.tnext - state.tprev
+
+            def history_extrapolation_implicit():
+                implicit_ts = state.dense_ts.at[state.dense_save_index].set(state.tnext)
```

Comment:
A solution would probably be that if we have an implicit step to make an in-place update of the `dense_ts` at the start of it all and then revert back to the old `dense_ts` ? (As you said that in-place updates are risky/dangerous, I have personally never done this before but see that the code does it here and there !) This is indeed an issue if `max_steps` is large since it would eat up the memory for the implicit step....